core/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java

```diff
@@ -391,7 +391,7 @@ public void finalizeRecovery() {
         StopWatch stopWatch = new StopWatch().start();
         logger.trace("[{}][{}] finalizing recovery to {}", indexName, shardId, request.targetNode());
         cancellableThreads.execute(() -> {
-            recoveryTarget.finalizeRecovery();
+            recoveryTarget.finalizeRecovery(shard.getGlobalCheckpoint());
             shard.markAllocationIdAsInSync(recoveryTarget.getTargetAllocationId());
```

Comment:
paranoia - can we flip this around and mark the target allocation as "in sync" before we give it the global checkpoint?  it at least reads better as "we know you are in sync and therefore every global checkpoint advances will take you into account"